### PR TITLE
First iteration of TraversabilityGrid, TraversabilityCell and TraversabilityClass.

### PR DESCRIPTION
**Pull request by JanWehrmann**
Contains basic data structures and a serialization test.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -3,6 +3,7 @@
 cmake_minimum_required(VERSION 2.6)
 set(ROCK_USE_CXX11 TRUE)
 find_package(Rock)
+set(ROCK_TEST_ENABLED OFF)
```
Comment (saarnold) [inline CMakeLists.txt]: This shouldn't be part of the PR.

File: src/grid/TraversabilityClass.cpp [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+{
+}
+
+TraversabilityClass::TraversabilityClass(): drivability(-1)
+{
+}
+
+bool TraversabilityClass::isSet() const
+{
+  return drivability < 0;
```
Comment (saarnold) [inline src/grid/TraversabilityClass.cpp]: The indentation seems to differ here. Please use everywhere (in this library) 4 spaces.

File: src/grid/TraversabilityClass.hpp [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+
+	/**
+	* Returns a value in the interval [0, 1].
+	* Zero means not drivable at all and
+	* one means perfect ground for driving.
+	* */
+	double getDrivability() const;
+	
+    private:
+	double drivability;
```
Comment (saarnold) [inline src/grid/TraversabilityClass.hpp]: A float should have enough precision for the drivability.

File: src/grid/TraversabilityGrid.hpp [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+      
+      // Set the traversabilityClassId and probability at the given coordinates.
+      // Note that probability has to be in range [0,1].
+      void setTraversabilityAndProbability(uint8_t traversabilityClassId, double probability, size_t x, size_t y);
+      
+      void setTraversability(uint8_t traversabilityClassId, size_t x, size_t y);
+      const TraversabilityClass& getTraversability(size_t x, size_t y) const;
+      
+      // Note that probability has to be in range [0,1].
+      void setProbability(double probability, size_t x, size_t y);
```
Comment (saarnold) [inline src/grid/TraversabilityGrid.hpp]: Since the probability is encoded in a uint8 a float would be enough here as well

File: src/grid/TraversabilityGrid.hpp [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+   * */
+  class TraversabilityGrid : public GridMap<TraversabilityCell> 
+  {    
+  public:
+      enum CLASSES {
+            CLASS_UNKNOWN = 0,
+            CLASS_OBSTACLE = 1,
+            CUSTOM_CLASSES = 2
+      };
+      CLASSES classes;
```
Comment (saarnold) [inline src/grid/TraversabilityGrid.hpp]: What is the reason for this member?

File: src/grid/TraversabilityGrid.hpp [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+   * (uncertainty of the data for this point). To conserve storage space both 
+   * values are stored in the grid as an uint8_t. For the traversability this 
+   * represents the Id of a TraversabilityClass in traversabilityClasses which 
+   * holds the actual value. The probability is simply converted from float [0, 1] 
+   * to uint8_t  [0, 255].
+   * */
+  class TraversabilityGrid : public GridMap<TraversabilityCell> 
+  {    
+  public:
+      enum CLASSES {
```
Comment (saarnold) [inline src/grid/TraversabilityGrid.hpp]: To align with the rest of the naming this should be CamelCase as well. Therefore just name it Classes.

File: src/grid/TraversabilityGrid.hpp (hunk 5 above)
Comment (JanWehrmann) [reply src/grid/TraversabilityGrid.hpp]: It made serialization of itself possible, which is indeed unnecessary.

File: src/grid/TraversabilityGrid.hpp [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+      CLASSES classes;
+	
+  private:
+      std::vector<TraversabilityClass> traversabilityClasses;
+          
+  public:
+      TraversabilityGrid();     
+      
+      // Use this if you want a map holding individual data.
+      TraversabilityGrid(const Vector2ui &num_cells, const Vector2d &resolution, const TraversabilityCell default_value);
```
Comment (saarnold) [inline src/grid/TraversabilityGrid.hpp]: default_value was supposed to be a reference right?

File: test/serialization/test_serialization_Grid.cpp [hunk 8]
```diff
@@ -231,6 +232,57 @@ BOOST_AUTO_TEST_CASE(test_mlgrid_serialization)
...
+	uint8_t returnId;
+	TraversabilityClass travClass = TraversabilityClass(0.1 * i);
+	BOOST_CHECK(traversabilityGrid_out.registerNewTraversabilityClass(returnId, travClass));
+    }
+    
+    for(unsigned i = 0; i < size.x(); i++)
+    {
+	for(unsigned j = 0; j < size.y(); j++)
+	{
+	    traversabilityGrid_out.setTraversabilityAndProbability(i % 4, j, i, j);
```
Comment (saarnold) [inline test/serialization/test_serialization_Grid.cpp]: For the probability only values in [0,1] are valid right?

File: src/grid/TraversabilityGrid.hpp (hunk 7 above)
Comment (JanWehrmann) [reply src/grid/TraversabilityGrid.hpp]: Yes!

File: test/serialization/test_serialization_Grid.cpp [hunk 9]
```diff
@@ -231,6 +232,57 @@ BOOST_AUTO_TEST_CASE(test_mlgrid_serialization)
...
+    boost::archive::binary_oarchive outarchive(stream);
+    outarchive << traversabilityGrid_out;
+    
+    boost::archive::binary_iarchive inarchchive(stream);
+    TraversabilityGrid traversabilityGrid_in = TraversabilityGrid();
+    inarchchive >> traversabilityGrid_in;
+    
+    std::vector<TraversabilityClass> classes_out = traversabilityGrid_out.getTraversabilityClasses();
+    std::vector<TraversabilityClass> classes_in = traversabilityGrid_in.getTraversabilityClasses();
+
```
Comment (saarnold) [inline test/serialization/test_serialization_Grid.cpp]: Here you should also check if the girds and classes have the same size.

**Comment by saarnold** [comment]
Ok looks good! Thanks!
